qldpc/codes.py

```diff
@@ -507,87 +657,94 @@ def get_code_params(
 
         Keyword arguments are passed to the calculation of code distance.
         """
-        distance = self.get_distance(pauli=None, lower=lower, upper=upper, **decoder_args)
+        distance = self.get_distance(
+            pauli=None, bound=bound, from_logical_ops=from_logical_ops, vector=None, **decoder_args
+        )
         return self.num_qudits, self.dimension, distance
 
     def get_distance(
         self,
         pauli: Literal[Pauli.X, Pauli.Z] | None = None,
         *,
-        lower: bool = False,
-        upper: int | None = None,
+        bound: int | None = None,
+        from_logical_ops: bool = False,
+        vector: npt.NDArray[np.int_] | None = None,
         **decoder_args: object,
     ) -> int:
-        """Distance of the this code: minimum weight of a nontrivial logical operator.
+        """Estimate the code distance, or the minimal weight of a nontrivial logical operator.
 
         If provided a Pauli as an argument, compute the minimim weight of an nontrivial logical
         operator of the corresponding type.  Otherwise, minimize over Pauli.X and Pauli.Z.
 
-        If `lower is True`, compute a lower bound: for the X-distance, compute the distance of the
-        classical Z-type subcode that corrects X-type errors.  Vice versa with the Z-distance.
+        If `bound is None`, compute an exact code distance by brute force.  Otherwise, compute an
+        upper bound using a randomized algorithm described in arXiv:2308.07915, minimizing over
+        `bound` random trials.  For a detailed explanation, see `CSSCode.get_one_distance_bound`.
 
-        If `upper is not None`, compute an upper bound using a randomized algorithm described in
-        arXiv:2308.07915, minimizing over `upper` random trials.  For a detailed explanation, see
-        `CSSCode.get_distance_upper_bound` and `CSSCode.get_one_distance_upper_bound`.
+        If `from_logical_ops is True`, estimate the code distance by minimizing the weight of
+        logical operators in a fixed basis, and identifying the lowest-weight operator.
 
-        If `lower is False` and `upper is None`, compute an exact code distance with integer linear
-        programming.  Warning: this is an NP-complete problem and takes exponential time to execute.
+        If provided a vector, compute the minimum Hamming distance between this vector and a
+        nontrivial X-type or Z-type logical operator, as applicable.
```

Comment:
Actually, is this correct?  In the case of the ClassicalCode, when provided a vector we compute its distance to *any* code word.  Should we be computing distance to *any* logical operator, or to a *nontrivial* logical operator here?

Reply:
I think as implemented it is nontrivial, but I don't think it matters unless the input vector in the dual (C_Z^\perp) space, which would never be an input. I am not even sure we actually used this vector generalization (for quantum codes)

Reply:
I guess one question is: what might this method be used for?  I struggle to think of an application for finding the distance between a vector and a nontrivial operator.  On the other hand, the distance between a vector and *any* logical operator has at least one use: it's the weight of a maximum-likelihood correction to an error vector.

If we can't think of any other applications, maybe we should (a) remove the `vector` option for quantum codes, or (b) change it to compute minimal distance to any logical operator (which has the benefit of parity with ClassicalCode behavior).

Thoughts?

Reply:
Actually, I just realized that even our `ClassicalCode` methods aren't self-consistent.  As an example, the following exact calculation:
```python
from qldpc import codes
code = codes.ClassicalCode.repetition(3)
print(code.get_distance(vector=[0, 0, 0]))
```
prints `0`, whereas
```python
print(code.get_distance(vector=[0, 0, 0], bound=1))
```
prints `3`.  The issue is that the first of these is computing distance from code words, whereas the second computes distance from nonzero code words.  Which of these do we want?

Reply:
Based on my memory of offline discussions, we want distance from nontrivial code words.  In that case we should fix the exact calculation in `ClassicalCode` (as well as the documentation).  I think this would leave `ClassicalCode` consistent with `CSSCode`, so it would be case closed?

Reply:
I guess if vector is in code (which we can easily check and handle) we do want to say the distance is 0 (so we want to fix get_distance_bound) (the condition b is not needed in this case as answer is just 0).
For a vector not in the code, the condition b check is anyway redundant as the zero vector does not have zero syndrome.  So basically, we need the following change -- if vector is not none, check if it is in the code, if so output 0. Else, do same computation without the part (b) check. Does that make sense?

Reply:
I think this makes sense to me.  So basically now `ClassicalCode.get_distance_bound` would use the non-commutation trick if `vector is None`, and otherwise just decode with `self.matrix @ vector` as the syndrome.  This is not unlike how `ClassicalCode.get_distance_bound` has explicitly different treatment for `vector is None` vs an all-0 vector.  Does that sound right?

Reply:
Yes, exactly

Reply:
Okay done.  I made `CSSCode` behave similarly.  Please take a look at the treatment of `vector` in the distance calculations in `ClassicalCode` and `CSSCode` to check consistency and correctness.